File: Python/Algorithm/Kadane_(Maximum_Subarray_Sum).py
# Complexity: O(n)
# Application: Used to determine the maximum subarray sum.

# To determine the summation:

def kad():
    mx , csum = ls[0], 0
    for e in ls:
        csum += e; mx = max(mx, csum); csum = max(csum, 0)
    return mx

# To determine the indices with summation:

def kad():
    msum, csum = float("-inf"), 0
    start = end = s = 0
    for i, num in enumerate(ls):
        csum += num
        if csum > msum: msum = csum; start, end = s, i
        if csum < 0: csum = 0; s = i + 1
    return start, end, msum


# Operation:
start, end, msum = kad(ls)

# Recursive Approach:

def kad(ls, left = None, right = None):
    if left is None and right is None:
        left, right = 0, len(ls)
    if left == right - 1: return ls[left]
    mid = (left + right) >> 1
    lmx = kad(ls, left, mid); rmx = kad(ls, mid, right)
    lsum = float('-inf'); rsum = float('-inf'); temp = 0
    for i in range(mid, right):
        temp += ls[i]; rsum = max(rsum, temp)
    temp = 0
    for i in range(mid - 1, left - 1, -1):
        temp += ls[i]; lsum = max(lsum, temp)
    return max(lsum + rsum, max(lmx, rmx))

File: CPP/Algorithm/Kadane_(Maximum_Subarray_Sum).cpp
// Complexity: O(n)
// To determine the summation:

int64_t mss(auto &aa) {
	int64_t n = aa.size(), mx = aa[0], res = 0;
	for (int i = 0; i < n; ++i) {
		res += aa[i]; mx = max(mx, res); res = max(res, 0ll);
	}
	return mx;
}

// To determine the indices of the range with the summation:
auto mss(auto &aa) {
	int64_t n = aa.size(), mx = aa[0], res = 0, start = 0, end = 0, s = 0;
	for (int i = 0; i < n; ++i) {
		res += aa[i];
		if (mx < res) { mx = res; start = s; end = i; }
		if (res < 0) { res = 0; s = i + 1; }
	}
	return make_pair(make_pair(start, end), mx);
}

// Recursive Approach:
int64_t aa[N], n;
int64_t mss() {
	if (n == 1) return aa[0];
	int64_t m = (n >> 1), lmx = mss(aa, m), rmx = mss(aa + m, n - m);
	int64_t lsum = LLONG_MIN, rsum = LLONG_MIN, sum = 0;
	for (int i = m; i < n; ++i) {
		sum += aa[i]; rsum = max(rsum, sum);
	}
	sum = 0;
	for (int i = m - 1; i >= 0; --i) {
		sum += aa[i]; lsum = max(lsum, sum);
	}
	return max(lsum + rsum, max(lmx, rmx));
}